clustermesh-apiserver/etcdinit/root.go

```diff
@@ -0,0 +1,242 @@
+		Info("Deleting contents of data directory")
+	// We don't use os.RemoveAll on the etcdDataDirectory because we don't want to remove the directory itself, just
+	// everything inside of it. In most cases that directory will be a mount anyway.
+	dir, err := os.ReadDir(etcdDataDir)
+	if err != nil {
+		log.WithField("etcdDataDir", etcdDataDir).
+			WithError(err).
+			Error("Failed to read from the etcd data directory while attempting to delete existing files")
+		return err
+	}
+	for _, d := range dir {
+		log.WithField("etcdDataDir", etcdDataDir).
+			WithField("path", d.Name()).
+			Debug("Removing file/directory in data dir")
+		err = os.RemoveAll(path.Join(etcdDataDir, d.Name()))
+		if err != nil {
+			log.WithField("etcdDataDir", etcdDataDir).
+				WithField("path", d.Name()).
+				WithError(err).
+				Error("Failed to remove pre-existing file/directory in etcd data directory")
+			return err
+		}
+	}
+
+	// Use "localhost" (instead of "http://127.0.0.1:2379" or "http://[::1]:2379") so it works in both the IPv4 and
+	// IPv6 cases.
+	loopbackEndpoint := "http://localhost:2379"
+	log.WithFields(logrus.Fields{
+		"etcdDataDir":             etcdDataDir,
+		"etcdListenClientUrl":     loopbackEndpoint,
+		"etcdClusterName":         etcdClusterName,
+		"etcdInitialClusterToken": etcdInitialClusterToken,
+	}).
+		Info("Starting localhost-only etcd process")
+	// Specify the full path to the etcd binary to avoid any PATH search binary replacement nonsense
+	etcdCmd := exec.CommandContext(ctx, etcdBinaryLocation,
+		fmt.Sprintf("--data-dir=%s", etcdDataDir),
+		fmt.Sprintf("--name=%s", etcdClusterName),
+		fmt.Sprintf("--listen-client-urls=%s", loopbackEndpoint),
+		fmt.Sprintf("--advertise-client-urls=%s", loopbackEndpoint),
+		fmt.Sprintf("--initial-cluster-token=%s", etcdInitialClusterToken),
+		"--initial-cluster-state=new")
+	log.WithField("etcdBinary", etcdBinaryLocation).
+		WithField("etcdFlags", etcdCmd.Args).
+		Debug("Executing etcd")
+
+	// Exec the etcd binary, which ultimately calls fork(2) under the hood. We don't wait on its completion, because
+	// it'll never complete of course.
+	err = etcdCmd.Start()
+	if err != nil {
+		log.WithField("etcdBinary", etcdBinaryLocation).
+			WithField("etcdFlags", etcdCmd.Args).
+			WithError(err).
+			Error("Failed to launch etcd process")
+		return err
+	}
+	etcdPid := etcdCmd.Process.Pid
+	log.WithField("etcdPID", etcdPid).
+		Info("Local etcd server process started")
+
+	// Defer etcd process cleanup
+	defer func() {
+		log := log.WithField("etcdPID", etcdPid)
+		log.Debug("Cleaning up etcd process")
+		// Send the process a SIGTERM. SIGTERM is the "gentle" shutdown signal, and etcd should close down it's resources
+		// cleanly and then exit.
+		log.Info("Sending SIGTERM signal to etcd process")
+		err := etcdCmd.Process.Signal(syscall.SIGTERM)
+		if err != nil {
+			log.WithError(err).
+				Error("Failed to send SIGTERM signal to etcd process")
+			returnErr = err
```

Comment:
I honestly didn't check all the codepaths (kudos to you for that), but yeah, I agree that it should never happen in practice. My point was only about the fact that in that case we would be failing anyhow, so it made sense to return immediately (and typically not returning after an error could be an just oversight, so it raises more attention). But I'm fine with either approach.

Reply:
I was wondering if sending a SIGTERM to etcd after it's been killed by context timeout would error and we'd end up in here instead of the message about timeout below, but that doesn't happen actually:
```bash
$ docker run -v ~/etcd-init-tests/:/var/run/etcd/ -it quay.io/cilium/clustermesh-apiserver:latest etcdinit --timeout=100ms
level=info msg="Starting first-time initialisation of etcd for Cilium Clustermesh" clusterName=default etcdClusterName=clustermesh-apiserver etcdDataDir=/var/run/etcd etcdInitialClusterToken=clustermesh-apiserver subsys=etcdinit timeout=100ms
level=info msg="Deleting contents of data directory" etcdDataDir=/var/run/etcd subsys=etcdinit
level=info msg="Starting localhost-only etcd process" etcdClusterName=clustermesh-apiserver etcdDataDir=/var/run/etcd etcdInitialClusterToken=clustermesh-apiserver etcdListenClientUrl="http://localhost:2379" subsys=etcdinit
level=info msg="Local etcd server process started" etcdPID=13 subsys=etcdinit
level=info msg="Starting etcd init" clusterName=default subsys=etcdinit
level=info msg="Performing pre-init tasks" subsys=etcdinit
{"level":"warn","ts":"2023-11-16T13:52:21.910574Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0x4000530380/localhost:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest balancer error: last connection error: connection error: desc = \"transport: Error while dialing: dial tcp 127.0.0.1:2379: connect: connection refused\""}
level=error msg="Failed to initialise etcd" clusterName=default error="setting key to indicate config is already set: setting path 'cilium/.has-cluster-config' to 'true': context deadline exceeded" subsys=etcdinit
level=info msg="Sending SIGTERM signal to etcd process" etcdPID=13 subsys=etcdinit
level=info msg="Waiting for etcd process to exit" etcdPID=13 subsys=etcdinit
level=error msg="etcd exited, but our context has expired. etcd may have been terminated due to timeout. Consider increasing the value of the timeout using the --timeout flag or CILIUM_TIMEOUT environment variable." etcdPID=13 subsys=etcdinit timeout=100ms
```
The SIGTERM goes off just fine once the context expires, even though we should kill it when that happens. I even put in a thread sleep to ensure it was done in time, and it still works fine. 🤔 

I'm not sure what to do here. I'm trying to avoid another run so I'll leave it for now, but I can change it in a follow-up if you like.